lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -1479,12 +1479,34 @@ static LogicalResult canonicalizeSingleSetConnect(StrictConnectOp op,
       return failure();
 
   } else {
-    // Constants/invalids in the same block are ok to forward, even through
+    // Values in the same block are ok to forward, even through
     // reg's since the clocking doesn't matter for constants.
-    if (!isa<ConstantOp>(srcValueOp) && !isa<InvalidValueOp>(srcValueOp))
-      return failure();
     if (srcValueOp->getBlock() != declBlock)
       return failure();
+
+    if (isa<ConstantOp, InvalidValueOp>(srcValueOp)) {
+      // Make sure the constant dominates all users.
+      srcValueOp->moveBefore(&declBlock->front());
+    } else if (isa<WireOp>(connectedDecl)) {
+      // If the decl is a wire, we can forward the value if the connection
+      // dominates the users.
+      // FIXME: There was a regression possibly related to `isBeforeInBlock`, so
```

Comment:
Can you expand?

Reply:
If there is a regression and we don't know why, we don't know why this might fix it.

Reply:
When I tried debugging https://github.com/llvm/circt/pull/3284, I found there was enormous number of calls to `isBeforeInBlock`. `isBeforeInBlock` is claimed to be (amortized) constant time but in the situation where we move the operations around, I thought `isBeforeInBlock` would be cause of the issue. I agree that I have to make sure things clear.